### PR TITLE
feat!: do not output files in editor

### LINKED ISSUE #93: FEATURE REQUEST: mapping via static call in-Editor
I'm a big fan of SceneReference but I never liked how it created json files for serialization inside a folder you need to source control. It keeps me from referencing this package inside of other packages, since I would have to communicate users they have to put these extra lines in their source control, and users may find my packages from multiple different sources.

What if you serialized guids inside a json file only once you build the game? And then, in the editor, it uses (something like) this?

```csharp
public class SceneReference
{
    public SceneAsset sceneAsset;

    public string path => AssetDatabase.GetAssetPath(sceneAsset);
}
```

Obviously wrapped in `#if UNITY_EDITOR` where appropriate.

### PR DESCRIPTION
**Pull request by starikcetin**
Closes #93

### CODE REVIEW

**Comment by starikcetin** [comment]
> Hi, I see this feature uses EditorPrefs. I'm worried that this might cause some weird results when you switch projects. (Note that EditorPrefs is shared between all of your Unity projects). Make sure that the json is regenerated once the editor loads.
> 
> This will however still be problematic if for whatever reason someone decides to have 2 Unity projects open at the same time. A better solution might be to use EditorPrefs with a key that is unique per-project.

Don't bother reviewing right now. It's a draft, it'll change and get cleaned up.

**Comment by starikcetin** [comment]
@CaseyHofland @mnemotic and everyone else interested, please review. I will merge in a week latest.

**Comment by mnemotic** [comment]
I don't feel I'm qualified to review your code, but, like @CaseyHofland suggested, consider including `Application.productName` in the `EditorPrefs` key to reduce the chance of any collisions?

**Comment by starikcetin** [comment]
> I don't feel I'm qualified to review your code, but, like @CaseyHofland suggested, consider including `Application.productName` in the `EditorPrefs` key to reduce the chance of any collisions?

We are not using EditorPrefs anymore, instead we are saving to UserSettings, which is already project specific.

**Comment by mnemotic** [comment]
Ah. Then my comment doesn't apply. Show's exactly how qualified I am to review this. 😁

**Comment by starikcetin** [comment]
> Ah. Then my comment doesn't apply. Shows exactly how qualified I am to review this. 😁

Don't be so hard on yourself :)

**Comment by starikcetin** [comment]
@CaseyHofland Do you have any comments? I am going to merge it soon.

File: Eflatun.SceneReference/Packages/com.eflatun.scenereference/Editor/SceneDataMapsGenerator.cs
Comment (CaseyHofland) [inline Eflatun.SceneReference/Packages/com.eflatun.scenereference/Editor/SceneDataMapsGenerator.cs:1]: Replacing `Run()` with `Run(bool)` will not be backwards compatible, and if I'm right the rest of this commit is completely backwards compatible. It is a shame to leave low-hanging fruit so I suggest providing an overload that will fill out `Run(bool)` with the most expected behaviour, e.g.

```csharp
public static void Run() => Run(default);
```

If it really is not supposed to be called like this anymore, then do the above but give it an `[ObseleteAttribute]` to ease conversion.

**Review by CaseyHofland — CHANGES_REQUESTED** [review]
Sorry, I kinda forgot about this! Usersettings is definitely the right solution, I don't use them enough myself!

There's one single change I would make because it's easy to do and would allow you to only do a minor increment.

File: Eflatun.SceneReference/Packages/com.eflatun.scenereference/Editor/SceneDataMapsGenerator.cs
Comment (starikcetin) [reply Eflatun.SceneReference/Packages/com.eflatun.scenereference/Editor/SceneDataMapsGenerator.cs:1]: As for backwards compatibility: Behaviour is part of public API. We changed the behaviour, therefore this PR is breaking, no matter the API changes.

Comment (CaseyHofland) [reply Eflatun.SceneReference/Packages/com.eflatun.scenereference/Editor/SceneDataMapsGenerator.cs:1]: Gotcha, well motivated. I still think letting the obsolete flag throw a warning is enough, although I have definitely worked in projects where a recompile would trigger 100+ warnings and the lead told me to “just clear it”; absolutely horrific.

**Review by CaseyHofland — APPROVED** [review]

File: Eflatun.SceneReference/Packages/com.eflatun.scenereference/Editor/SceneDataMapsGenerator.cs
Comment (CaseyHofland) [reply Eflatun.SceneReference/Packages/com.eflatun.scenereference/Editor/SceneDataMapsGenerator.cs:1]: I don't think I have permission to resolve this conversation btw, may that be correct?

**Comment by starikcetin** [comment]
Will be included in the next release.

**Comment by CaseyHofland** [comment]
Chapeau

I have to personally thank you for the quick turnaround, when I asked for this feature I never expected it would get picked up so quickly. You’re an amazing creator you know that?

See you later space cowboy

**Comment by starikcetin** [comment]
> Congrats!
> 
> I have to personally thank you for the quick turnaround, when I asked for this feature I never expected it would get picked up so quickly. You’re an amazing creator you know that?
> 
> See you space cowboy

It is very nice to hear kind words like this, thank you.

**Pull request merged**